rosbag2_py/src/rosbag2_py/_writer.cpp

```diff
@@ -104,28 +83,39 @@ std::unordered_set<std::string> get_registered_serializers()
 
 }  // namespace rosbag2_py
 
+using PyWriter = rosbag2_py::Writer<rosbag2_cpp::writers::SequentialWriter>;
+using PyCompressionWriter = rosbag2_py::Writer<rosbag2_compression::SequentialCompressionWriter>;
+
 PYBIND11_MODULE(_writer, m) {
   m.doc() = "Python wrapper of the rosbag2_cpp writer API";
 
-  pybind11::class_<rosbag2_py::Writer<rosbag2_cpp::writers::SequentialWriter>>(
-    m, "SequentialWriter")
-  .def(pybind11::init())
-  .def("open", &rosbag2_py::Writer<rosbag2_cpp::writers::SequentialWriter>::open)
-  .def("write", &rosbag2_py::Writer<rosbag2_cpp::writers::SequentialWriter>::write)
-  .def("remove_topic", &rosbag2_py::Writer<rosbag2_cpp::writers::SequentialWriter>::remove_topic)
-  .def("create_topic", &rosbag2_py::Writer<rosbag2_cpp::writers::SequentialWriter>::create_topic);
-
-  pybind11::class_<rosbag2_py::Writer<rosbag2_compression::SequentialCompressionWriter>>(
-    m, "SequentialCompressionWriter")
+  pybind11::class_<PyWriter>(m, "SequentialWriter")
   .def(pybind11::init())
-  .def("open", &rosbag2_py::Writer<rosbag2_compression::SequentialCompressionWriter>::open)
-  .def("write", &rosbag2_py::Writer<rosbag2_compression::SequentialCompressionWriter>::write)
   .def(
-    "remove_topic",
-    &rosbag2_py::Writer<rosbag2_compression::SequentialCompressionWriter>::remove_topic)
+    "open",
+    pybind11::overload_cast<
+      const rosbag2_storage::StorageOptions &, const rosbag2_cpp::ConverterOptions &
+    >(&PyWriter::open))
```

Comment:
@emersonknapp Out of curiosity why do we need `pybind11::overload_cast<..>` for open method here?

Reply:
It's because `rosbag2_cpp::Writer` has two overloaded `::open` methods. (`open(const std::string & uri)` and `open(const StorageOptions &, const ConverterOptions &)`) Python doesn't allow for function overloading, so we have to specify which of the two signatures to expose as the `.open` method here. Without the `overload_cast`, compilation fails due to the ambiguous signature. 

We didn't need this before because we had defined `rosbag2_py::Writer::open` method explicitly. Now, by changing to inherit from `rosbag2_cpp::Writer` instead, both `open` methods are available and we have to specify which one we mean.

Reply:
@emersonknapp Thanks for explanation. Now it became clear for me.